sw/device/lib/dif/dif_padctrl.c

```diff
@@ -0,0 +1,305 @@
+  }
+
+  register_index = index / PADCTRL_MIO_PADS_ATTR_FIELDS_PER_REG;
+
+  return base_register_offset + (kPadctrlRegisterWidthInBytes * register_index);
+}
+
+/**
+ * Returns true if padctrl is locked, assumes `padctrl` argument is valid
+ */
+static bool padctrl_is_locked(const dif_padctrl_t *padctrl) {
+  return !mmio_region_get_bit32(padctrl->base_addr, PADCTRL_REGEN_REG_OFFSET,
+                                PADCTRL_REGEN_WEN);
+}
+
+/**
+ * Pad change types for use with `change_attrs`
+ */
+typedef enum attr_change_type {
+  kAttrChangeEnable,  /**< Enable attributes that have bit set. */
+  kAttrChangeDisable, /**< Disable attributes that have bit set. */
+  kAttrChangeWrite    /**< Overwrite field with provided bits. */
+} attr_change_type_t;
+
+/**
+ * Change attributes for a particular pad.
+ *
+ * @param padctrl Padctrl state data.
+ * @param kind What kind of pad is being changed.
+ * @param index Which pad is being changed.
+ * @param attrs Attribute bits that specify the change.
+ * @param change_type What change is applied using the `change_attrs`.
+ * @return dif_padctrl_change_attr_result_t.
+ */
+static dif_padctrl_change_attr_result_t change_attrs(
+    const dif_padctrl_t *padctrl, dif_padctrl_pad_kind_t kind,
+    dif_padctrl_pad_index_t index, dif_padctrl_attr_t attrs,
+    attr_change_type_t change_type) {
+  // Check arguments before doing anything.
+  if (padctrl == NULL) {
+    return kDifPadctrlChangeBadArg;
+  }
+
+  if (!is_pad_index_valid(kind, index)) {
+    return kDifPadctrlChangeBadArg;
+  }
+
+  if (!attrs_are_valid(attrs)) {
+    return kDifPadctrlChangeBadArg;
+  }
+
+  // Cannot continue with locked padctrl as registers cannot be altered.
+  if (padctrl_is_locked(padctrl)) {
+    return kDifPadctrlChangeLocked;
+  }
+
+  // Calculate which register and which field within it relate to the pad that
+  // is the target of the attribute change.
+  bitfield_field32_t attr_field = calc_pad_attr_field(kind, index);
+  uint32_t attr_reg_offset = calc_pad_attr_reg_offset(kind, index);
+
+  // Read current attribute register and extract relevant field.
+  uint32_t attr_reg_value =
+      mmio_region_read32(padctrl->base_addr, attr_reg_offset);
+
+  attr_field.value = bitfield_get_field32(attr_reg_value, attr_field);
+
+  // Alter field depending upon change type.
+  switch (change_type) {
+    case kAttrChangeEnable:
+      attr_field.value |= attrs;
+      break;
+    case kAttrChangeDisable:
+      attr_field.value &= ~attrs;
+      break;
+    case kAttrChangeWrite:
+      attr_field.value = attrs;
+      break;
+    default:
+      return kDifPadctrlBadArg;
+  }
+
+  // Check if the new setting causes a conflict.
+  if (conflicting_attributes(attr_field.value)) {
+    return kDifPadctrlChangeAttrConflict;
+  }
+
+  // Write back updated value to attribute register.
+  uint32_t new_attr_reg_value =
+      bitfield_set_field32(attr_reg_value, attr_field);
+
```

Comment:
Delete this newline, to merge the two statements that make the write into a single stanza.